src/openms/source/FILTERING/ID/IDFilter.cpp

```diff
@@ -782,6 +782,34 @@ namespace OpenMS
     }
   }
 
+  void IDFilter::keepNBestSpectra(std::vector<PeptideIdentification>& spectra, Size n)
+  {
+    for (PeptideIdentification& s : spectra) { s.sort(); }
+
+    // there might be less spectra identified then n -> adapt
+    n = std::min(n, spectra.size());    
+    std::partial_sort(spectra.begin(), spectra.begin() + n, spectra.end(),
+      [] (const PeptideIdentification& l, const PeptideIdentification& r) 
+      {
+        if (r.getScoreType() != l.getScoreType())
```

Comment:
this compares nlogn times, but only n times would be required.
How about
`if (find_if(spectra.begin(), spectra.end(), [&](const PeptideIdentification& pep) { return (spectra.begin()->getScoreType() !=pep.getScoreType()) } != spectra.end()) throw ...` before the sorting
`for_each` would be even less code, but we're not at C++17 **yet**

Reply:
It’s Approximately (last-first)log(middle-first) applications of cmp. So it is basically O(n) in practice. Don’t know if worth the effort

Reply:
overhead aside, it just felt like a surprising place to find this checking code inside a sorting subfunction...
I have no strong feelings here, but it's a minor "WTF" moment.

Reply:
ok I will do it in an external loop